cpp/src/arrow/compute/kernels/match_test.cc

```diff
@@ -0,0 +1,457 @@
+
+  // No match
+  CheckMatch<FixedSizeBinaryType, std::string>(
+      &this->ctx_, fixed_size_binary(5),
+      /* in_values= */ {"bbbbc", "bbbbc", "aaaad", "cccca"},
+      /* in_is_valid= */ {},
+      /* member_set_values= */ {"bbbbb", "", "bbbbb", "aaaaa", "ddddd"},
+      /* member_set_is_valid= */ {true, false, true, true, true},
+      /* out_values= */ {0, 0, 0, 0},
+      /* out_is_valid= */ {false, false, false, false});
+
+  // Empty left array
+  CheckMatch<FixedSizeBinaryType, std::string>(&this->ctx_, fixed_size_binary(5), {}, {},
+                                               {"bbbbb", "", "bbbbb", "aaaaa", "ccccc"},
+                                               {true, false, true, true, true}, {}, {});
+
+  // Empty right array
+  CheckMatch<FixedSizeBinaryType, std::string>(
+      &this->ctx_, fixed_size_binary(5), {"bbbbb", "", "bbbbb", "aaaaa", "ccccc"},
+      {true, false, true, true, true}, {}, {}, {0, 0, 0, 0, 0},
+      {false, false, false, false, false});
+
+  // Empty arrays
+  CheckMatch<FixedSizeBinaryType, std::string>(&this->ctx_, fixed_size_binary(0), {}, {},
+                                               {}, {}, {}, {});
+}
+
+TEST_F(TestMatchKernel, MatchDecimal) {
+  std::vector<Decimal128> input{12, 12, 11, 12};
+  std::vector<Decimal128> member_set{12, 12, 11, 12};
+  std::vector<int32_t> expected{0, 1, 2, 0};
+
+  CheckMatch<Decimal128Type, Decimal128>(&this->ctx_, decimal(2, 0), input,
+                                         {true, false, true, true}, member_set,
+                                         {true, false, true, true}, expected, {});
+}
+
+TEST_F(TestMatchKernel, MatchChunkedArrayInvoke) {
+  std::vector<std::string> values1 = {"foo", "bar", "foo"};
+  std::vector<std::string> values2 = {"bar", "baz", "quuux", "foo"};
+  std::vector<std::string> values3 = {"foo", "bar", "foo"};
+  std::vector<std::string> values4 = {"bar", "baz", "barr", "foo"};
+
+  auto type = utf8();
+  auto a1 = _MakeArray<StringType, std::string>(type, values1, {});
+  auto a2 = _MakeArray<StringType, std::string>(type, values2, {true, true, true, false});
+  auto a3 = _MakeArray<StringType, std::string>(type, values3, {});
+  auto a4 = _MakeArray<StringType, std::string>(type, values4, {});
+
+  ArrayVector array1 = {a1, a2};
+  auto carr = std::make_shared<ChunkedArray>(array1);
+  ArrayVector array2 = {a3, a4};
+  auto member_set = std::make_shared<ChunkedArray>(array2);
+
+  auto i1 = _MakeArray<Int32Type, int32_t>(int32(), {0, 1, 0}, {});
+  auto i2 =
+      _MakeArray<Int32Type, int32_t>(int32(), {1, 2, 2, 2}, {true, true, false, false});
+
+  ArrayVector expected = {i1, i2};
+  auto expected_carr = std::make_shared<ChunkedArray>(expected);
+
+  Datum encoded_out;
+  ASSERT_OK(Match(&this->ctx_, carr, member_set, &encoded_out));
+  ASSERT_EQ(Datum::CHUNKED_ARRAY, encoded_out.kind());
+
+  AssertChunkedEqual(*expected_carr, *encoded_out.chunked_array());
+}
+
```

Comment:
Please add tests against larger, randomly generated inputs (at least for the basic numeric types).